tests/clpy_tests/core_tests/test_ndarray_adv_indexing.py

```diff
@@ -371,8 +371,8 @@ def test_adv_setitem(self, xp, dtype):
      'value': numpy.array([1, 2, 3, 4])},
     {'shape': (2, 3, 4), 'indexes': (slice(None), [0, -1]),
      'value': numpy.arange(2 * 2 * 4).reshape(2, 2, 4)},
-    {'shape': (2, 3, 4), 'indexes': (slice(None), [[0, 1], [2, 0]]),
-     'value': numpy.arange(2 * 2 * 2 * 4).reshape(2, 2, 2, 4)},
+    # {'shape': (2, 3, 4), 'indexes': (slice(None), [[0, 1], [2, 0]]),
```

Comment:
Can you explain me as following in detail?
* What this disable test case does
* What `double assignment` means on your description
* Why CuPy is able to pass (and ClPy is not)

Reply:
To put it simply,
this parameterization does like:
```
a = numpy.zeros(3)
a[[0, 1, 2, 0]] = [10, 20, 30, 40]
```
through [`_scatter_update_kernel`](https://github.com/fixstars/clpy/pull/196/files#diff-351cb10d746e6f305386b5a499a06e26R2773).

There are four simultaneous assignments:
```
a[0] = 10
a[1] = 20
a[2] = 30
a[0] = 40
```
The result `a` can be `[10, 20, 30]` and `[40, 20, 30]` depending on the assignment order.
Numpy's document says that there no guarantee of the order of assignments:
> For advanced assignments, there is in general no guarantee for the iteration order. This means that if an element is set more than once, it is not possible to predict the final result.
> Cited from: https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html

I couldn't find out why CuPy accepts this case.

Reply:
The interested test case actually does

```
a = clpy.zeros((2, 3, 4), dtype='int32')
a[(slice(None), [[0, 1], [2, 0]])] = numpy.arange(2 * 2 * 2 * 4).reshape(2, 2, 2, 4).astype(a.dtype)
```

Reply:
In summary, 

```
src=[10, 20, 30, 40]
dst=[0, 0, 0, 0]
dst[2] = src[2]
dst[1] = src[1]
dst[0] = src[0] or src[4] (data race????)
```

The test expects latter but NumPy does not guarantee it (could be former). Therefore we can not test it with NumPy. right?

The problems are
* ClPy must be compatible with CuPy, not NumPy directly. Therefore, ClPy must return `[40, 20, 30]` as its return value if CuPy guarantees it.
* If the returned value by NumPy is detemined randomly because of its unguarantee, we must test it with other way. (This means the CuPy's test is also in the same situation thus we should send PR them).

@nsakabe-fixstars please investigate it

Reply:
Your understanding is right.

#### 1st item
I finally found a section mentioning the difference of the behaviour of `ndarray.__setitem__`.
> CuPy’s `__setitem__` behaves differently from NumPy when integer arrays reference the same location multiple times. In that case, the value that is actually stored is undefined. Here is an example of CuPy.
> NumPy stores the value corresponding to the last element among elements referencing duplicate locations.
https://docs-cupy.chainer.org/en/stable/reference/difference.html?highlight=indexing#duplicate-values-in-indices

CuPy doesn't guarantee the assignment order (undefined).


#### 2nd item
NumPy's assignment results are not random as I said above.
It deterministically overwrites `dst[0]` by `src[3]` (symbols are from your example).
On the other hand, CuPy's result is _undefined_.

I executed corresponding CuPy's test case on our machine. It failed by 100%:

`TestArrayAdvancedIndexingVectorValue`
* CuPy v2.1.0 on _titanv_
    * fails
* CuPy v5.3.0(latest stable) on _titanv_
    * fails

In my opinion, both we and CuPy's developers have to exclude this test because the behavior totally depends on the environment.

Reply:
> In my opinion, both we and CuPy's developers have to exclude this test (parameter) because the behavior totally depends on the environment.

How about latest CuPy? Is this excluded (if so, when?) or stil fail?

Reply:
> CuPy v5.3.0(latest stable) on titanv
>     fails

You mean the test fails on the latest? If so, we must report it to CuPy team.

Reply:
Yes, the interested test case, `TestArrayAdvancedIndexingVectorValue(param_3)` is included and doesn't pass in `CuPy v5.3.0(latest stable)`.

```
test_ndarray_adv_indexing.py ................................................................... [ 25%]
................................................................................................ [ 61%]
..............................................................................F................. [ 97%]
.......                                                                                          [100%]

=============================================== FAILURES ===============================================
____________________ TestArrayAdvancedIndexingVectorValue_param_3.test_adv_setitem _____________________
(omitted)
```

Do I send a PR them?

Reply:
I confirmed the fail. As we discuss orally, please send PR to remove the test for CuPy team.

Reply:
I sent a pr them https://github.com/cupy/cupy/pull/2088 alongside leaving a note here.

Reply:
Nice! I approve to remove this test case